src/DocumentDB.ChangeFeedProcessor/FeedProcessing/ChangeFeedQueryTimeoutDecorator.cs

```diff
@@ -0,0 +1,49 @@
+﻿// ----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+// ----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor.FeedProcessing
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.DataAccess;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Monitoring;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Utils;
+    using Microsoft.Azure.Documents.Client;
+
+    internal class ChangeFeedQueryTimeoutDecorator : IChangeFeedDocumentQuery<Document>
+    {
+        private readonly IChangeFeedDocumentQuery<Document> query;
+        private readonly IHealthMonitor monitor;
+        private readonly TimeSpan timeout;
+
+        public ChangeFeedQueryTimeoutDecorator(IChangeFeedDocumentQuery<Document> query, IHealthMonitor monitor, TimeSpan timeout)
+        {
+            this.query = query;
+            this.monitor = monitor;
+            this.timeout = timeout;
+        }
+
+        public bool HasMoreResults => this.query.HasMoreResults;
+
+        public async Task<IFeedResponse<TResult>> ExecuteNextAsync<TResult>(CancellationToken token = default(CancellationToken))
+        {
+            try
+            {
+                return await this.query.ExecuteNextAsync<TResult>(token).AbortOnTimeout(this.timeout, token).ConfigureAwait(false);
+            }
+            catch (TimeoutException e)
```

Comment:
I think the issue wasn't about handling timeout exception (PartitionProcessor would already handle that), I thought it was about detecting a hang in the ExecuteNextAsync call, so that we call it and it never returns -- in this case we need to  pull cancellation token or do something else if that doesn't help, essentially we need to detect the hang and abort the call. Is that right?

Reply:
Noticed the AbortOnTimeout. Not sure if that's enough though. Can we launch a parallel task at the same level where we get the task for ExecuteNextAsync? Potentilly ExectureNextAsync may hang even before it retunrs